# Compare AM and EM abundances with tree abundances for manuscript figure 5 

library(tidyverse)
library(data.table)
library(ggpmisc)
library(ggpubr)

# EM/AM tree abundance - file from supplement of Lang et al. 2023
AM_trees = read_csv("reference_data/AM_trees/AM_trees_lang.csv")

# Read in metagenomic abundances from Braken
bracken_genus = fread("data/classification/taxonomic_rank_summaries/genus/soil_microbe_db_filtered_genus_merged_lineage.csv", nThread = 8)

# Read in funguild assignments per-genus
# Generated by scripts/summarize_outputs/summarize_its_genus_funguild.r
funguild_results_full <- fread("data/classification/analysis_files/FUNguild_assignments.csv")

# Create list of EMF/AMF taxa
funguild_results_simple = funguild_results_full  %>% filter(grepl("Prob", confidenceRanking)) %>% 
    select(name, guild) %>% 
    mutate(Ectomycorrhizae = ifelse(grepl("Ecto", guild), T, F),
           "Arbuscular mycorrhizae" = ifelse(grepl("Arbusc", guild), T, F),
           "Pathogen" = ifelse(grepl("Pathogen", guild), T, F),
           "Endophyte" = ifelse(grepl("Endophyte", guild), T, F),
           "Saprotroph" = ifelse(grepl("Sapro", guild), T, F)) %>% 
    select(-guild)
emf_list <- funguild_results_simple %>% filter(Ectomycorrhizae==T) %>% select(name) %>% unlist
amf_list <- funguild_results_simple %>% filter(`Arbuscular mycorrhizae`==T) %>% select(name) %>% unlist


# Now use EMF/AMF lists to assign taxon status
lineage_df =  lapply(unique(bracken_genus$lineage),
                     function(x) {
                         phyloseq::parse_taxonomy_qiime(x) %>% 
                             t()  %>% as.data.frame()
                     }) %>%
    data.table::rbindlist()
lineage_df = cbind.data.frame(lineage = unique(bracken_genus$lineage), 
                              lineage_df)
bracken_genus = left_join(bracken_genus, lineage_df)

fungal_phyla = c("Ascomycota","Basidiomycota","Blastocladiomycota",
                 "Chytridiomycota","Cryptomycota","Mucoromycota",
                 "Microsporidia","Olpidiomycota","Zoopagomycota")

bracken_genus$is_fungi = ifelse(bracken_genus$Phylum %in% fungal_phyla, T, F)
bracken_genus$is_emf = ifelse(tolower(bracken_genus$Genus) %in% emf_list, T, F)
bracken_genus$is_amf = ifelse(tolower(bracken_genus$Genus) %in% amf_list, T, F)

# Ugly code below

# Calculate relative abundance of EMF compared to fungal abundance
emf_counts_rel = bracken_genus %>% group_by(sample_id, is_fungi) %>% 
    mutate(fungi_abun = sum(fraction_total_reads)) %>% ungroup %>% 
    group_by(sample_id,is_fungi, fungi_abun, is_emf) %>% 
    summarize(emf_abun = sum(fraction_total_reads)) %>% ungroup %>% 
    filter(is_emf==T & is_fungi==T) %>% select(-c(is_emf, is_fungi)) %>% 
    mutate(emf_abun_rel = emf_abun/fungi_abun)

emf_counts_rel$db_name = "soil_microbe_db"
emf_counts_rel = emf_counts_rel %>% 
    separate(sample_id,  
             into = c("compositeSampleID","db_name"), 
             sep = "COMP_", remove = F, extra = "merge") %>% 
    mutate(compositeSampleID = str_remove(sample_id, paste0("_",db_name))) 
emf_counts_rel = 
    cbind(emf_counts_rel,
          microbialForecast:::parseNEONsampleIDs(emf_counts_rel$compositeSampleID) %>% 
              select(siteID, plotID, dateID, horizon, plot_date)) %>% 
    mutate(Site = siteID)
emf_counts_rel$`Plot Number` = substr(emf_counts_rel$plotID, 6, 8) %>% as.numeric(gsub("^0|^00", "", .))


emf_rel_EM_trees = merge(emf_counts_rel, AM_trees)

# Figure 5c
p2 = ggplot(emf_rel_EM_trees,
            aes(x = `Proportion ECM basal area`, y = emf_abun_rel)) + 
    geom_jitter(alpha=.8, size=2, width = .01, height=.01) + 
    stat_smooth(method="lm") +
    scale_y_log10() + 
    theme_bw(base_size = 16)  +
    ylab("EMF rel.abundance in metagenome")  +
    xlab("EMF-associated tree basal area") + 
    stat_cor(label.y.npc = 1, size=7, p.accuracy = .0001) +
    stat_regline_equation(aes(label = ..rr.label..),
                          show.legend = FALSE, size=7, label.y.npc = .95) +  
    theme(plot.margin = unit(c(3, 25, 25, 45), "pt"),
          axis.title = element_text(face = "bold"))
p2

# Save figure 5c
ggsave("manuscript_figures/fig5c.png", p2, width = 10, height = 8, units = "in", dpi = 300)



# Repeat for AMF!
# Calculate relative abundance of AMF compared to fungal abundance
amf_counts_rel = bracken_genus %>% group_by(sample_id, is_fungi) %>% 
    mutate(fungi_abun = sum(fraction_total_reads)) %>% ungroup %>% 
    group_by(sample_id,is_fungi, fungi_abun, is_amf) %>% 
    summarize(amf_abun = sum(fraction_total_reads)) %>% ungroup %>% 
    filter(is_amf==T & is_fungi==T) %>% select(-c(is_amf, is_fungi)) %>% 
    mutate(amf_abun_rel = amf_abun/fungi_abun)

amf_counts_rel$db_name = "soil_microbe_db"
amf_counts_rel = amf_counts_rel %>% 
    separate(sample_id,  
             into = c("compositeSampleID","db_name"), 
             sep = "COMP_", remove = F, extra = "merge") %>% 
    mutate(compositeSampleID = str_remove(sample_id, paste0("_",db_name))) 
amf_counts_rel = 
    cbind(amf_counts_rel,
          microbialForecast:::parseNEONsampleIDs(amf_counts_rel$compositeSampleID) %>% 
              select(siteID, plotID, dateID, horizon, plot_date)) %>% 
    mutate(Site = siteID)
amf_counts_rel$`Plot Number` = substr(amf_counts_rel$plotID, 6, 8) %>% as.numeric(gsub("^0|^00", "", .))



amf_rel_AM_trees = merge(amf_counts_rel, AM_trees)

# Figure 5b
p1 = ggplot(amf_rel_AM_trees,
            aes(x = `Proportion AM basal area`, y = amf_abun_rel)) + 
    geom_jitter(alpha=.8, size=2, width = .01, height=.01) + 
    stat_smooth(method="lm") +
    scale_y_log10() + 
    theme_bw(base_size = 16)  +
    ylab("AMF rel.abundance in metagenome")  +
    xlab("AMF-associated tree basal area") + 
    stat_cor(label.y.npc = 1, size=7, p.accuracy = .0001) +
    stat_regline_equation(aes(label = ..rr.label..),
                      show.legend = FALSE, size=7, label.y.npc = .95) +  
    theme(plot.margin = unit(c(3, 25, 25, 45), "pt"),
          axis.title = element_text(face = "bold")) 

# Save figure 5b
ggsave("manuscript_figures/fig5b.png", p1, width = 10, height = 8, units = "in", dpi = 300)

# Save figure 5c
ggsave("manuscript_figures/fig5c.png", p2, width = 10, height = 8, units = "in", dpi = 300)

# Figure 5 (combined)
# Note: fig_5a is generated by scripts/visualize/compare_its.r
# If fig_5a doesn't exist, source compare_its.r to generate it
if(!exists("fig_5a")) {
    cat("⚠️  fig_5a not found in environment. Running scripts/visualize/compare_its.r to generate it...\n")
    source("scripts/visualize/compare_its.r")
}

if(exists("fig_5a")) {
    fig_5bc = ggarrange(p1, p2, common.legend = T, nrow = 1, labels = c("B","C"))
    fig5 = ggarrange(fig_5a, fig_5bc, nrow=2, labels = c("A","",""),
                     heights = c(2,1)) 
    ggsave("manuscript_figures/fig5.png", fig5, width = 12, height = 14, units = "in", dpi = 300)
    cat("✅ Saved combined figure to: manuscript_figures/fig5.png\n")
} else {
    cat("⚠️  Error: Could not generate fig_5a. Please check scripts/visualize/compare_its.r\n")
}
