# Calculate the % classified at the species level across databases, biomes, and seq depth
library(tidyverse)
library(data.table)
library(ggpubr)
library(ggpmisc)
library(rstatix)
source("scripts/helper_functions.r")
# Note: source.R from comets_shinyapp_example may not be available locally
# source("/projectnb/talbot-lab-data/zrwerbin/soil_microbe_GEMs/comets_shinyapp_example/source.R")

# generated by scripts/summarize_outputs/summarize_pct_filtered.r
pass_filter_species_long = read_csv("data/classification/analysis_files/pass_filter_summary.csv")


# Get list of samples that have been evaluated by all databases
pass_filter_species_long$common =  gsub("-COMP","",pass_filter_species_long$sampleID)
sample_count = pass_filter_species_long %>% group_by(common) %>% tally
common_samples = sample_count[sample_count$n>2,]$common


# Check for significant differences between databases
db_fit <- pass_filter_species_long  %>% 
    filter(common %in% common_samples) %>%  
    filter(metric=="percent_passing") %>% 
    group_by(pretty_metric) %>% 
    anova_test(value ~ db_name) %>% 
    add_significance()

#### Run Tukey ###
db_tukey <- pass_filter_species_long %>% filter(common %in% common_samples) %>%  
    group_by(pretty_metric) %>% 
    tukey_hsd(value ~ db_name) %>% 
    add_significance() %>% 
    add_xy_position(step.increase = .01)

db_tukey$y.position=sqrt(db_tukey$y.position)

# Visualize classification between databases
fig2 = pass_filter_species_long  %>% 
    filter(common %in% common_samples) %>%  
    filter(metric=="percent_passing") %>% 
    ggplot(aes(y = value, x=reorder(db_name, -value), color=db_name)) +
    geom_boxplot(outlier.shape = NA, show.legend = F) +
    geom_point(#aes(color=fungal),
        size=2, alpha=.4, show.legend = F,
        position=position_jitter(width = .1, height=0.01)) + 
    theme_bw(base_size = 18) + 
    ylab("% metagenomic reads classified to species") + 
    xlab(NULL) +
    guides(color=guide_legend(NULL)) +
    stat_compare_means(show.legend = F)  + 
    scale_y_sqrt() + #scale_y_sqrt() +
    stat_pvalue_manual(db_tukey %>%  
                           filter(pretty_metric=="% reads classified at\n high quality"),#y.position = .21,
                       #bracket.nudge.y = .34,
                       hide.ns = T, y.position = .3,
                       bracket.nudge.y = .2, step.increase = .05)#+ ggtitle("Classified to the species level") + ylim(0, .25)
fig2

# Save figure
ggsave("manuscript_figures/fig2.png", fig2, width = 10, height = 8, units = "in", dpi = 300)

# Print stats for % classified and kept after filtering
pass_filter_species_long %>% group_by(metric, db_name) %>% 
    filter(common %in% common_samples) %>% 
    summarize(mean = mean(value, na.rm=T), median = median(value, na.rm=T))



# Now add in biome information for Figure 5 - only SMDB 
# Load soilCores using helper function
if(!exists("soilCores")) {
    soilCores <- load_soilCores()
}
if(exists("soilCores")) {
soilData_subset = soilCores %>% 
    filter(compositeSampleID %in% pass_filter_species_long$sampleID) %>% 
    select(compositeSampleID, siteID, biome, horizon) %>% 
    distinct(compositeSampleID, .keep_all = T) %>% 
    filter(!is.na(compositeSampleID))

pass_filter_soilCore = left_join(pass_filter_species_long %>%
                                     mutate(compositeSampleID = sampleID), 
                                 soilData_subset)  %>% 
    filter(sampleID != "TALL_004-M-20140708-COMP" ) # this sample got corrupted at some point
} else {
    warning("⚠️  MISSING FILE: soilCores not found!")
    warning("   Continuing without biome information - biome analysis will be skipped")
    # If soilCores is not available, use pass_filter_species_long directly
    pass_filter_soilCore = pass_filter_species_long %>%
        mutate(compositeSampleID = sampleID,
               siteID = substr(sampleID, 1, 4),
               biome = NA,
               horizon = NA)
}


# Subset for plotting
species_pass_filter = pass_filter_soilCore %>% 
    filter(db_name =="soil_microbe_db") %>% 
    filter(metric == "percent_passing") %>% 
    filter(seq_depth > 100000) 

species_pass_filter$isWetland = ifelse(species_pass_filter$biome=="Wetlands", T,F)

species_pass_filter$biome = factor(species_pass_filter$biome, levels = c("Shrubland","Herbaceous","Forest", "Cultivated","Wetlands"))

# Check for significant differences between biomes (only if biome data available)
if(!all(is.na(species_pass_filter$biome))) {
fit <- species_pass_filter %>% 
    group_by(db_name, pretty_metric) %>% 
    anova_test(value ~ biome) %>% 
    add_significance()
#### Run Tukey ###
tukey <- species_pass_filter %>% 
    group_by(db_name, pretty_metric) %>% 
    tukey_hsd(value ~ biome) %>% 
    add_significance() %>% 
    add_xy_position(step.increase = .01)

tukey$y.position=sqrt(tukey$y.position)
} else {
    # Skip biome analysis if biome data is not available
    species_pass_filter$biome = factor(rep("Unknown", nrow(species_pass_filter)))
    fit = NULL
    tukey = NULL
}


c = ggplot(species_pass_filter,
         #  aes(y = value, x=reorder(biome, -value), color=biome)) +
    aes(y = value, x=biome, color=biome)) +
    geom_boxplot(outlier.shape = NA, show.legend = F) +
    geom_point(size=2, alpha=.4, show.legend = F,
        position=position_jitter(width = .3, height=0.01)) + 
    theme_bw(base_size = 18) + 
    ylab("% reads classified to species at\n high quality") + 
    xlab(NULL) +
    guides(color=guide_legend(NULL)) +
    stat_compare_means(show.legend = F)  + 
    scale_y_sqrt()  + #scale_y_sqrt() +
    {if(!is.null(tukey)) stat_pvalue_manual(tukey,#y.position = .21,
                       #bracket.nudge.y = .34,
                       hide.ns = T, y.position = .34,
                       bracket.nudge.y = .22, step.increase = .01)}
c


# Now look at sequencing depth
b <- ggplot(species_pass_filter,
       aes(y = value, x=seq_depth)) +
    geom_point(#aes(color=fungal),
        size=2, alpha=.4, show.legend = F,
        position=position_jitter(width = .1, height=0.01)) + 
    theme_bw(base_size = 18) + 
    ylab("% reads classified to species at\n high quality") + 
    stat_smooth(method = "loess") +
    xlab(NULL) +
    guides(color=guide_legend(NULL)) +
    scale_y_log10() + 
    #scale_x_log10() + 
    scale_x_continuous(
        trans = "log10",
        breaks=c(1e5,1e6,2.5e6,5e6,1e7,5e7,1e8),
        labels = c("100K","1M","2.5M","5M","10M","50M","100M") )


# Proportion fungi vs sequencing depth
phyla_fungi_summary <- read_csv("data/classification/analysis_files/soil_microbe_db_phyla_fungi_summary.csv")
fungi_proportion <- left_join(species_pass_filter, phyla_fungi_summary)

a <- ggplot(fungi_proportion,
       aes(y = phyla_fungi_metagenome, x=seq_depth)) +
    geom_point(#aes(color=fungal),
        size=2, alpha=.4, show.legend = F,
        position=position_jitter(width = .1, height=0.01)) + 
    theme_bw(base_size = 18) + 
    ylab("Fungal proportion in \nmetagenome") + 
    xlab("Sample sequencing depth (read count)") +
    stat_smooth(method = "gam", formula = y ~ s(x, bs = "cs")) +
    #stat_smooth(method = "loess") +
    
    guides(color=guide_legend(NULL)) +
    scale_y_log10() + 
    #tagger::tag_facets(tag_levels = "A")  +
    scale_x_continuous(
        trans = "log10",
        #trans="sqrt",
        breaks=c(1e5,1e6,2.5e6,5e6,1e7,5e7,1e8),labels = c("100K","1M","2.5M","5M","10M","50M","100M") ) +
    
    stat_cor(aes(label=paste(..rr.label.., ..p.label.., sep = "~`,`~")), 
             # method="pearson", 
             label.y = 0, size=7)

left_panel=ggarrange(a, b, nrow=2, labels = c("A","B"))

left_panel

ggarrange(left_panel, c, nrow=1, labels = c(NA,"C"))
