# Compare metagenomic abundances with those from ITS rRNA sequencing

library(tidyverse)
library(data.table)
library(ggpubr)
library(ggupset)
library(correlation)
library(patchwork)
library(ggstatsplot)
library(ggrepel)

# Generated by /scripts/summarize_outputs/summarize_by_its_genus_and_funguild.r
df_compare <- read_csv("/projectnb/frpmars/soil_microbe_db/data/NEON_metagenome_classification/summary_files/genus_ITS_metagenome_comparison.csv")


# Generated by /scripts/summarize_outputs/summarize_by_its_genus_and_funguild.r
funguild_results_full <- fread("/projectnb/frpmars/soil_microbe_db/data/NEON_metagenome_classification/summary_files/FUNguild_assignments.csv")
df_compare <- left_join(df_compare, funguild_results_full) 

# Visualize a handful of taxa for manuscript
ggplot(df_compare %>% filter(name %in% c("cenococcum","glomus","russula","umbelopsis","suillus","penicillium")), 
       aes(x = metagenome_abundance, y = ITS_abundance)) + 
    geom_point(alpha=.5)  +
    stat_smooth(method="lm") +
    scale_y_sqrt() + 
    scale_x_sqrt() + 
    geom_abline(slope=1, intercept = 0, linetype=2) + 
    stat_regline_equation(aes(label = ..rr.label..),
                          show.legend = FALSE, size=6, label.x.npc = .4) + 
    facet_grid(~name, scales="free") + 
    theme_bw(base_size = 18)  +
    xlab("Relative abundance in metagenome") +
    ylab("Relative abundance in ITS amplicon")  + 
    theme(axis.text.x = element_text(angle = 270))


summary_df = df_compare %>% 
    drop_na(metagenome_abundance, ITS_abundance) %>% 
    #rename(a = metagenome_abundance, b = ITS_abundance) %>% 
    mutate(a = sqrt(metagenome_abundance), b = sqrt(ITS_abundance)) %>% 
    group_by(legacy, name) %>% 
    add_tally() %>% 
    filter(n > 5) %>% 
    summarize(R = cor.test(a, b,method = "spearman")$estimate, p = cor.test(a, b,method = "spearman")$p.value) %>%
    mutate(R = as.numeric(formatC(R, format = "fg", digits = 3)),
           p = as.numeric(formatC(p, format = "g", digits = 3))) %>%
    drop_na(R) %>% 
    mutate(Rsq = R^2,
           significant = ifelse(p < .05, 21, 19))
    #gather(key = measure, value = value, -exp) %>%
    #unite("stat", measure, value, sep = " = ")



funguild_upset = funguild_results_full  %>% filter(grepl("Prob", confidenceRanking)) %>% 
                    select(name, guild) %>% 
                    mutate(EMF = ifelse(grepl("Ecto", guild), T, F),
                                                  "AMF" = ifelse(grepl("Arbusc", guild), T, F),
                                                  "Pathogen" = ifelse(grepl("Pathogen", guild), T, F),
                                                  "Endophyte" = ifelse(grepl("Endophyte", guild), T, F),
                           "Saprotroph" = ifelse(grepl("Sapro", guild), T, F)
                           #"Other" = ifelse((!Ectomycorrhizae & !Pathogen & !Saprotroph & !`Arbuscular mycorrhizae`), T, F)
                    ) %>% 
    select(-guild)

                                     
funguild_upset_tidy <- funguild_upset %>%
    as_tibble() %>%
    gather(guild, Member, -name) %>%
    filter(Member) %>%
    select(- Member) %>%
    group_by(name) %>%
    mutate(AMF = ifelse("Arbuscular mycorrhizae" %in% guild, "AMF", "Not AMF")) %>% 
    group_by(name, AMF) %>% 
    summarize(guild = list(guild))
#%>%

guild_scores <- inner_join(funguild_upset_tidy, summary_df) %>% filter(legacy==F)


most_abundant = df_compare %>% group_by(name) %>% 
    summarize(median = median(metagenome_abundance, na.rm=T)) %>%
    arrange(-median) %>% 
    head(., 100)
#filter(median > .005)

guild_scores <- guild_scores %>% distinct(name, AMF, .keep_all=TRUE) %>% 
    mutate(guild_collapsed = sapply(guild, function(x) paste0(sort(x), collapse = "-")))



intersections_to_keep = table(guild_scores[guild_scores$name %in% most_abundant$name,]$guild_collapsed) %>% sort %>% tail(8) %>%  names

# Create labels for values that are negative and significant, or positive and strong, or AMF
labels <- guild_scores %>% 
    filter(guild_collapsed %in% intersections_to_keep) %>% 
    
    filter(name %in% most_abundant$name & 
               ((R < 0 & p < .05) | (R >.3 & p < .05) ) | name %in% c("glomus","rhizophagus"))



fig_5a = guild_scores %>%
    mutate(guild_collapsed = factor(guild_collapsed, levels = c("Pathogen",
                                                                 "Endophyte-Pathogen-Saprotroph",
                                                                 "Saprotroph",
                                                                "Pathogen-Saprotroph",
                                                                "Endophyte-Saprotroph",
                                                                "AMF-Endophyte",
                                                                 "EMF-Saprotroph",
                                                                 "EMF"), ordered=T)) %>% 

    filter(guild_collapsed %in% intersections_to_keep) %>% 
    
            filter(name %in% most_abundant$name) %>% 
    ggplot(aes(x = guild_collapsed, y = R)) +
    geom_violin(draw_quantiles = c(.5)) + 
    geom_point(aes(#color=guild_collapsed, #alpha = abs(log(p))), 
                   shape=as.factor(significant)),
               position = position_jitter(height=0, width=.01), alpha=.7, size=3) +
    ylab("Spearman Ï between rel. abundance \nin ITS and metagenome")  +
    axis_combmatrix(sep="-") +
    theme_bw(base_size = 18) + 
    scale_shape_manual(values = c(21, 19), 
                       labels = c("Not significant", "Significant"))  +
    theme_combmatrix(combmatrix.label.text = element_text(color = "black", 
                                                          size=18),
                     combmatrix.label.make_space = F,
                     plot.margin = unit(c(1.5, 35, 10, 45), "pt")) + 
    theme(legend.position=c("top"),
          axis.text.y = element_text(size=14),
          legend.title=element_blank(),
          axis.title = element_text(face = "bold")) + 
    xlab("Functional guild")  #+ 
    #plot_annotation(tag_levels = list(c('A'), '1')) 
    # geom_label_repel(data=guild_scores %>%
    #                      filter(name %in% most_abundant$name) %>%
    #                      filter(R < -.1 & p < .05),
    #           aes(y = R, x = guild, label = name))   +
fig_5a = fig_5a + 
    geom_label_repel(aes(label = janitor::make_clean_names(name, case = "big_camel"), y = R), label.padding = .2,box.padding = .5,
                     data = labels, size=5)
    # geom_label_repel(data=guild_scores %>% 
    #                      filter(name %in% most_abundant$name) %>% 
    #                      filter(R > .45 & p < .05), 
    #                 aes(y = R, x = guild, label = name), 
    #                 min.segment.length = 0.01)
fig_5a

# Now run "compare_ecto_tree_abundance.r" to generate the other two panels of figure 5 (tree comparison)

#fig_5a %>% 
#ggexport(height = 1000, width = 1200, filename = "/projectnb/frpmars/soil_microbe_db/manuscript_figures/fig4.png")

